### PR TITLE
Add a Testing Guide

### PR DESCRIPTION
**Pull request by infiniteregrets**

### CODE REVIEW

**Comment by aviramha** [comment]
I think we should split it to setup a kubernetes environment so an user can understand they can use any k8s cluster they want (with the caveat they'll need to push the built image to be available there).
Formatting needs to be improved - more sections, some snippets issues etc.

File: content/docs/developer/testing/index.md [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+draft: false
+images: []
+menu:
+  docs:
+    parent: "testing"
+weight: 110
+toc: true
+---
+
+## Setup Environment
```
Comment (aviramha) [inline content/docs/developer/testing/index.md:15]: Start with listing the pre-requisites (Rust version, docker, kubernetes cluster [local/or remote])

Comment (aviramha) [reply content/docs/developer/testing/index.md:15]: point to a guide how to setup a kubernetes cluster (the guide you wrote)

File: content/docs/developer/testing/index.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+}
+2022-06-30T05:17:31.877655Z DEBUG mirrord_layer::tcp: handle_incoming_message -> handled Ok(
+    (),
+)
+2022-06-30T05:17:31.878193Z  WARN mirrord_layer::tcp_mirror: tcp_tunnel -> exiting due to remote stream closed!
+2022-06-30T05:17:31.878255Z DEBUG mirrord_layer::tcp_mirror: tcp_tunnel -> exiting
+OK - GET: Request completed
+```
+
+### Run E2E tests
```
Comment (aviramha) [inline content/docs/developer/testing/index.md:248]: Upper section

**Review by aviramha — CHANGES_REQUESTED** [review]

File: content/docs/developer/testing/index.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+
+### Prerequisites
+
+- [Rust](https://www.rust-lang.org/)
+- [Nodejs](https://nodejs.org/en/) & [Expressjs](https://expressjs.com/)
+- [Python](https://www.python.org/) & [Flask](https://flask.palletsprojects.com/en/2.1.x/)
+- Kubernetes Cluster (local/remote)
+
+### Setup a k8s Cluster
```
Comment (eyalb181) [inline content/docs/developer/testing/index.md:23]: Should be Kubernetes (for consistency's sake)

File: content/docs/developer/testing/index.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+A minimal Kubernetes cluster can be easily setup locally using either of the following -
+- [Minikube](https://minikube.sigs.k8s.io/)
```
Comment (eyalb181) [inline content/docs/developer/testing/index.md:26]: A. [Kind](https://kind.sigs.k8s.io/docs/user/quick-start/) is also a good option
B. I think our default is back to being Minikube? @aviramha

File: content/docs/developer/testing/index.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+### Preparing a cluster
+
+Switch Kubernetes context to `your-context`
+
+```bash
+kubectl config get-contexts
+```
+
+```bash
+kubectl config use-context your-context
```
Comment (eyalb181) [inline content/docs/developer/testing/index.md:49]: I thought we were conforming to docker-desktop (or minikube) for the entire guide?

File: content/docs/developer/testing/index.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+- Kubernetes Cluster (local/remote)
+
+### Setup a k8s Cluster
+
+A minimal Kubernetes cluster can be easily setup locally using either of the following -
+- [Minikube](https://minikube.sigs.k8s.io/)
+- [Docker Desktop](https://www.docker.com/products/docker-desktop/)
+
+For the ease of illustration and testing, we will conform to using Docker Desktop for the rest of the guide.
+### Docker Desktop
```
Comment (eyalb181) [inline content/docs/developer/testing/index.md:30]: Make sure to sort out the heading hierarchy here. Docker Desktop and Preparing a cluster should be subheadings of Setup a Kubernetes cluster

File: content/docs/developer/testing/index.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+py-serv      NodePort    10.96.139.36   <none>        80:32095/TCP   3h8m
+❯ kubectl get deployments
+NAME                 READY   UP-TO-DATE   AVAILABLE   AGE
+py-serv-deployment   1/1     1            1           3h8m
+❯ kubectl get pods
+NAME                                 READY   STATUS    RESTARTS   AGE
+py-serv-deployment-ff89b5974-x9tjx   1/1     Running   0          3h8m
+```
+
+### Build mirrord-agent
```
Comment (eyalb181) [inline content/docs/developer/testing/index.md:142]: I would specify "Build mirrord-agent Docker image", and add a note as to why we're not building the actual Rust library

File: content/docs/developer/testing/index.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+docker build -t test . --file mirrord-agent/Dockerfile
+```
+
+```bash
+❯ docker images
+REPOSITORY                                     TAG       IMAGE ID       CREATED         SIZE
+test                                           latest    5080c20a8222   2 hours ago     300MB
+```
+
+### Build & run mirrord
```
Comment (eyalb181) [inline content/docs/developer/testing/index.md:154]: Build and Run mirrord*

File: content/docs/developer/testing/index.md [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+
+On Nightly toolchain -
```
Comment (eyalb181) [inline content/docs/developer/testing/index.md:156]: I would remove this sentence, it's confusing and evident from the next line anyway

File: content/docs/developer/testing/index.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+
+| OSX | `cargo +nightly build --workspace --exclude mirrord-agent` |
+| - | - |
+| **Linux** | **`cargo +nighty build`** |
+
+Run mirrord with a local process
+
+Sample web server - `app.js`
+
+```js
```
Comment (eyalb181) [inline content/docs/developer/testing/index.md:166]: Why is app.yaml collapsible but app.js isn't?

File: content/docs/developer/testing/index.md [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+2022-06-30T05:17:31.878255Z DEBUG mirrord_layer::tcp_mirror: tcp_tunnel -> exiting
+OK - GET: Request completed
+```
+
+### Run E2E tests
+
+Make sure `mirrord-agent` image is tagged as `test`
+
+```bash
+docker tag agent-image test
```
Comment (eyalb181) [inline content/docs/developer/testing/index.md:263]: Is this necessary? Isn't it created as 'test' given the build command you provided?

**Review by eyalb181 — CHANGES_REQUESTED** [review]

File: content/docs/developer/testing/index.md (hunk 4 above)
Comment (infiniteregrets) [reply content/docs/developer/testing/index.md:26]: Local testing was honestly quite easy with docker, didnt have to configure much. But I believe we should include docs for minikube and docker both. Lmk!

**Comment by aviramha** [comment]
Stale for too long, let's merge this then re-iterate later on.

**Pull request merged**